stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+            }
+            Err(_) => {
+                return Ok(false);
+            }
+        };
+
+        neighbor_rpc.send_request(network, self.naddr.clone(), request)?;
+        self.idle = false;
+        Ok(true)
+    }
+
+    /// Handle a received StacksHttpResponse and advance the state machine.
+    /// If we get the full tenure's blocks, then return them.
+    /// Returns Ok(Some([blocks])) if we successfully complete the state machine.
+    /// Returns Ok(None) if we accepted the response and did a state-transition, but we're not done
+    /// yet.  The caller should now call `send_next_download_request()`
+    /// Returns Err(..) on failure to process the response.
+    pub fn handle_next_download_response(
+        &mut self,
+        response: StacksHttpResponse,
+    ) -> Result<Option<Vec<NakamotoBlock>>, NetError> {
+        self.idle = true;
+        match self.state {
+            NakamotoTenureDownloadState::GetTenureStartBlock(_block_id) => {
+                test_debug!(
+                    "Got download response for tenure-start block {}",
+                    &_block_id
+                );
+                let block = response.decode_nakamoto_block()?;
+                self.try_accept_tenure_start_block(block)?;
+                Ok(None)
+            }
+            NakamotoTenureDownloadState::WaitForTenureEndBlock(..) => {
+                test_debug!("Invalid state -- Got download response for WaitForTenureBlock");
+                Err(NetError::InvalidState)
+            }
+            NakamotoTenureDownloadState::GetTenureEndBlock(_block_id) => {
+                test_debug!("Got download response to tenure-end block {}", &_block_id);
+                let block = response.decode_nakamoto_block()?;
+                self.try_accept_tenure_end_block(&block)?;
+                Ok(None)
+            }
+            NakamotoTenureDownloadState::GetTenureBlocks(_end_block_id) => {
+                test_debug!(
+                    "Got download response for tenure blocks ending at {}",
+                    &_end_block_id
+                );
+                let blocks = response.decode_nakamoto_tenure()?;
+                self.try_accept_tenure_blocks(blocks)
+            }
+            NakamotoTenureDownloadState::Done => Err(NetError::InvalidState),
+        }
+    }
+
+    pub fn is_done(&self) -> bool {
+        self.state == NakamotoTenureDownloadState::Done
+    }
+}
+
+/// Download states for a unconfirmed tenures.  These include the ongoing tenure, as well as the
+/// last complete tenure whose tenure-end block hash has not yet been written to the burnchain (but
+/// the tenure-start hash has -- it was done so in the block-commit for the ongoing tenure).
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) enum NakamotoUnconfirmedDownloadState {
+    /// Getting the tenure tip information
+    GetTenureInfo,
+    /// Get the tenure start block for the ongoing tenure
+    GetTenureStartBlock(StacksBlockId),
```

Comment:
I assume the StacksBlockId within GetTenureStartBlock is the StacksBlockId of the start block, but I am not sure what StacksBlockId is for in GetUnconfirmedTenureBlocks...Might be worth a comment? Granted I am not 100% sure on a lot of stuff here and I don't think that is the fault of the documentation XD

Reply:
Added